nexus/src/app/crucible.rs

```diff
@@ -0,0 +1,1086 @@
+                                    Err(Error::invalid_request(&rv.message))
+                                }
+
+                                _ => {
+                                    Err(Error::internal_error(&rv.message))
+                                }
+                            }
+                        }
+
+                        _ => {
+                            Err(Error::internal_error(
+                                "unexpected failure during `region_delete`",
+                            ))
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    /// Send a running snapshot deletion request
+    async fn request_crucible_running_snapshot_delete(
+        &self,
+        log: &Logger,
+        dataset: &db::model::Dataset,
+        region_id: Uuid,
+        snapshot_id: Uuid,
+    ) -> Result<(), Error> {
+        let client = self.crucible_agent_client_for_dataset(dataset);
+        let dataset_id = dataset.id();
+
+        let result = ProgenitorOperationRetry::new(
+            || async {
+                client
+                    .region_delete_running_snapshot(
+                        &RegionId(region_id.to_string()),
+                        &snapshot_id.to_string(),
+                    )
+                    .await
+            },
+            || async { self.crucible_agent_gone_check(dataset_id).await },
+        )
+        .run(log)
+        .await;
+
+        match result {
+            Ok(_) => Ok(()),
+
+            Err(e) => {
+                error!(
+                    log,
+                    "region_delete_running_snapshot saw {:?}",
+                    e;
+                    "dataset_id" => %dataset_id,
+                    "region_id" => %region_id,
+                    "snapshot_id" => %snapshot_id,
+                );
+
+                match e {
+                    ProgenitorOperationRetryError::Gone => {
+                        // Return Ok if the dataset's agent is gone, no delete
+                        // call is required.
+                        Ok(())
+                    }
+
+                    ProgenitorOperationRetryError::GoneCheckError(_) => {
+                        Err(Error::internal_error(
+                            "insufficient permission for crucible_agent_gone_check"
+                        ))
+                    }
+
+                    ProgenitorOperationRetryError::ProgenitorError(e) => match e {
+                        crucible_agent_client::Error::ErrorResponse(rv) => {
+                            match rv.status() {
+                                status if status.is_client_error() => {
+                                    Err(Error::invalid_request(&rv.message))
+                                }
+
+                                _ => {
+                                    Err(Error::internal_error(&rv.message))
+                                }
+                            }
```

Comment:
similarly, this could be:nit, take it or leave  it: maybe:
```suggestion
                            if rv.status().is_client_error() => {
                                Err(Error::invalid_request(&rv.message))
                            } else {
                                Err(Error::internal_error(&rv.message))
                            }
```